WIP Facade GUI merger

thecodewarrior opened the pull request:
## Necessary before merging
- [x] ~~Bitfont needs to be removed, as it adds several megabytes of dependency in the form of ICU.~~ Turns out file size isn't an issue. Betweenlands is 80mb and Astral Sorcery is 24, so an extra 12mb is no big deal
- [x] All Vec3d/Vec2d/BlockPos constructor usages should use the caches.
- [x] Add options to sprite mcmeta for +/- justification on each axis (i.e. options so a clipped sprite's right or bottom edge is preserved and the left or top are cropped respectively)
- [x] Remove GuiWindow and convert the color picker window to a component
- [x] Find entry points of old GUI system and make sure they are wired up to the new one too
- [x] Rename to Facade
- [x] Fix mouseOver not respecting parent's clipping
- [x] Add `@Experimental` to Bitfont and its components
- [ ] Remove defunct holdover components
- [x] Add sprite rotation
- [ ] Add IMValue caching (optionally cache value per frame with `refreshRate` and `refresh()`)
- [x] Better scroll events
- [x] Improve `StackLayout` 
  - [x] make `StackLayout` an interface and have `StackLayoutLayer` and `StackLayoutComponent`

## Non-critical todo
- [x] Improve pastry tests
- [x] Fix pastry dropdown extending off the screen (truncate and scroll)
- [x] add option to `StackLayout` to preserve crosswise positioning

## Pastry todo
- [ ] Figure out how to convey which control is focused
- [x] Button
  - [ ] Enabled/Disabled appearance
- [x] Checkbox
  - [ ] Enabled/Disabled appearance
- [x] Radio button
  - [ ] Enabled/Disabled appearance
- [x] Toggle switch
  - [ ] Enabled/Disabled appearance
- [x] Checkbox/radio button set
  - [ ] needs API improvements
- [x] Color picker
- [x] Progress bar
  - [x] Basic implementation
  - [ ] Vertical option
  - [ ] Investigate API improvements
- [ ] Text editor
  - [x] Basic implementation
  - [ ] Polish
  - [ ] Improve API _drastically_ (the api is virtually non-existent)
    - [ ] Research other text editor APIs
  - [ ] Enabled/Disabled appearance
- [x] Dropdown
  - [x] Basic implementation
  - [x] Scroll when list extends offscreen
  - [ ] Enabled/Disabled appearance
- [x] Tooltip
  - [ ] Time-based one?
  - [ ] Help ones that can be turned off and on?
- [ ] Pastry help system?
  - [ ] Add pastry indicator that tells people the goal of pastry (people might not know and still expect it to behave like a normal MC interface)
  - [ ] Add system to turn on/off help tooltips or a help overlay?
- [ ] Context menu (right-click menu. Not sure whether straying so far from MC may make stuff _more_ confusing)
- [x] Scroll pane
- [x] Scrollbar
  - [x] Basic implementation
  - [x] Variable handle height
- [ ] Themed standard colors in `PastryTexture` (background, linework, text, etc.)
- [x] Tab pane
  - [ ] Scroll tabs when too large
  - [ ] Allow stretching tabs to fill width
- [ ] Slider
  - [ ] Basic implementation
  - [ ] Snapping
  - [ ] Tick marks
  - [ ] Handle type and orientation
  - [ ] Selection (blue on one side)
  - [ ] range selection (two handles)
  - [ ] Enabled/Disabled appearance
- [ ] PastryBackground inner corners (maybe have `ComplexPastryBackground` which takes the frames of special sublayers and creates a background that covers that area using CTM or something of the like)
- [ ] Resizable split pane?
- [ ] Drag and drop
- [ ] Custom non-text clipboard

thecodewarrior commented:
About how experimental stuff is:
- Facade will be slightly experimental. I won't be tiptoeing around breaking changes, but I won't go doing massive refactors for fun. Facade won't have any `@Experimental` annotations however, because that will make using and developing it a royal pain in the ass
- Pastry will be a bit more experimental. Again I won't tiptoe around breaking changes, but again I won't go doing massive refactors for fun. It has pretty much the same policy, but is much less mature and so I will be breaking it a lot more. Unless it will negatively affect  the design I'll try to avoid flipping things on their head in ways that will make updating a pain though. Pastry won't have any `@Experimental` annotations either because, again, it will be a PITA to use and develop.
- The much more volatile APIs in Pastry (e.g. `PastryTextEditor` and `PastrySlider`) will have an `@Experimental` annotation however, `@VolatilePastryAPI`

SaadArdati commented on src/example/java/com/teamwizardry/librarianlib/test/facade/tests/GuiTestAnimationAwait.kt:
```diff
@@ -0,0 +1,52 @@
+        launch {
+            var i = 0
+            while(true) {
+                val x = ThreadLocalRandom.current().nextDouble(0.0, 100.0)
```
Could you not

thecodewarrior replied:
I took out the coroutine stuff because the ASM broke and it runs a high risk of memory leaks

SaadArdati commented on src/example/java/com/teamwizardry/librarianlib/test/facade/tests/GuiTestAnimationAwait.kt:
```diff
@@ -0,0 +1,52 @@
+
+    fun randomString(count: Int): String {
+        val source = "ABCDEFGHIJKLMNOPQRSTUVWXYZ\n"
+        return ThreadLocalRandom.current()
```
Is ThreadLocalRandom really necessary here?

thecodewarrior replied:
I took out the coroutine stuff because the ASM broke and it runs a high risk of memory leaks

thecodewarrior commented:
Just ran a binary compatibility report, and [:oof:](http://static.thecodewarrior.dev/LibrarianLib-1.19.1-neogui_compat_report.html).

Forgive the version naming on the page, this was just thrown together. `4.18-fat` is Facade, while `1.19.1...` is the current build on CurseForge. I’ve put a summary here so you don’t have to dredge through the entire page and so I have a checklist to tick things off on.

## Incompatibilities
- [x] Animation builder-style methods were somehow removed.
- [x] Somehow my shader stuff snuck its way in when I tried to pare down liblibx, so a ton of the shader stuff is gone.
- [x] When replacing primitive overloads with inline `Number` functions I was an idiot and removed the primitive versions.
- [x] I broke the existing `ComponentSpriteTiled` because I hadn’t implemented the pinning required to replace it.
- [x] As part of my sprite updates I added new non-default methods to `ISprite`. Oops.
- [x] As part of my sprite updates I removed the existing draw methods from `ISprite` and changed `DrawingUtil`. Double oops.

None of these should be a significant issue. 
- The animation methods can just be added back.
- The shader stuff can be reverted.
- The primitive overloads can be added back if I can figure out how to make Kotlin not prefer them (will it prefer non-deprecated overloads? I could just rename them all and use `@JvmName` to make them compile to the same bytecode).
- The tiled sprite can now be properly implemented.
- The new methods can be given defaults.
- The draw methods can probably be emulated without too much issue.

Bluexin commented:
Make sure to merge branch 1.12 into yours first.
That should already fix some of the missing stuff (looking at you animator builders) ;)

thecodewarrior commented:
So a few incompatibilities still exist. Firstly some protected fields in `ComponentSpriteTiled` were removed, but the biggest change is in `Sprite`. Most of the changes shouldn't effect anyone who isn't doing a lot of manual sprite manipulation though, since all it changes is making sprite no longer mutable (e.g. you can't just go and set their width or height any more) and removing the `getSubSprite` method which based on github search nobody uses anyway. It was really only used for `ComponentSpriteTiled`.

Bluexin commented:
How about deprecating their mutability ? Unless there is a technical reason to have them  mutable, in which case, let's discuss it.
Did you check to make sure `ComponentSpriteTiled` still works like before though ?
That `getSubSprite` method can be deprecated. That is, if there is an actual reason to remove it (does it hurt anything?)

thecodewarrior commented:
The issue is that sprites have changed quite a bit internally. `ISprite`s haven't, but the `Sprite`s generated by `Texture` have. Their implementation is no longer conducive to that kind of mutability without some messy hacks.
